### ISSUE TITLE
关于demo

### ISSUE THREAD

**Issue by Einkazusa** [tjiiv-cprg/EPro-PnP#78]
您好！首先非常感谢您在EPro-PnP上的研究
我想将EPro-PnP层加入自己的位姿测量网络中，在demo文件夹的fit_identity.ipynb中我看到了相关的提示。感谢您的帮助！我的网络是通过提取预定义3D点在RGB图像中映射的2D点，构成2D-3D匹配进行PnP解算，因此我对demo中所给的示例有些疑问：
1. 您以out_pose = EProPnP(MLP(in_pose))作为示例，这个MLP(in_pose)的作用是否等同于一个位姿测量网络的特征提取部分？由输入提取出的特征如x3d，x2d，w2d这些信息。但在我的网络中x3d是已知的真实数据，还需要作为参数输入model.forward_train中吗？
2. 同样是在fit_identity.ipynb中，out_pose由in_pose加上随机噪声得到，在model.forward_train和计算loss中都用到了out_pose，这里的out_pose相当于是监督训练用到的真实坐标数据吗？我理解的w2d是权重参数，在训练中out_pose中的w2d是必须的吗？因为在我的网络中并不需要w2d作为监督
如果能得到您的解答，我将不胜感激！

**Comment by Lakonik** [comment]
Hi! Thank you for your interest in our work.

1. Yeah these are dummy inputs (only to showcase the basic usage of the code). If you have the predefined keypoints then x3d should be the coordinates of those keypoints. For w2d, the only hard requirement is to apply softmax and global scaling after some initial predictions (the global scaling should be handled with care as explained in the updated arxiv 2023 preprint).

2. Yes, `out_pose` is the g.t. pose. During training you need to add g.t. as a candidate proposal to stabilize Monte Carlo sampling.

**Comment by Einkazusa** [comment]
Thank you for your response!

In addition, I have some further questions about a segment of code in the initialization of the demo:
# Here we use static weight_scale because the data noise is homoscedastic
self.log_weight_scale = nn.Parameter(torch.zeros(2))

Later in the code, there are parts related to self.log_weight_scale:
w2d = (w2d.log_softmax(dim=-2) + self.log_weight_scale).exp()
norm_factor = model.log_weight_scale.detach().exp().mean()

In this context, self.log_weight_scale is not updated through norm_factor, but it is updated through w2d. In practical usage, should .detach() be removed from self.log_weight_scale to compute the loss using norm_factor for updating instead of relying on w2d?

**Comment by Einkazusa** [comment]
Thank you for your response!

In addition, I have some further questions about a segment of code in the initialization of the demo:

# Here we use static weight_scale because the data noise is homoscedastic
self.log_weight_scale = nn.Parameter(torch.zeros(2))

Later in the code, there are parts related to self.log_weight_scale:
w2d = (w2d.log_softmax(dim=-2) + self.log_weight_scale).exp()
norm_factor = model.log_weight_scale.detach().exp().mean()

In this context, self.log_weight_scale is not updated through norm_factor, but it is updated through w2d. In practical usage, should .detach() be removed from self.log_weight_scale to compute the loss using norm_factor for updating instead of relying on w2d?

**Comment by Einkazusa** [comment]
Thank you for your response!

In addition, I have some further questions about a segment of code in the initialization of the demo:

## Here we use static weight_scale because the data noise is homoscedastic
self.log_weight_scale = nn.Parameter(torch.zeros(2))

Later in the code, there are parts related to self.log_weight_scale:
w2d = (w2d.log_softmax(dim=-2) + self.log_weight_scale).exp()
norm_factor = model.log_weight_scale.detach().exp().mean()

In this context, self.log_weight_scale is not updated through norm_factor, but it is updated through w2d. In practical usage, should .detach() be removed from self.log_weight_scale to compute the loss using norm_factor for updating instead of relying on w2d?

**Comment by Einkazusa** [comment]
- Thank you for your response!

In addition, I have some further questions about a segment of code in the initialization of the demo:

```python

# Here we use static weight_scale because the data noise is homoscedastic
self.log_weight_scale = nn.Parameter(torch.zeros(2))
```



Later in the code, there are parts related to `self.log_weight_scale`:

```python

w2d = (w2d.log_softmax(dim=-2) + self.log_weight_scale).exp()
norm_factor = model.log_weight_scale.detach().exp().mean()
```



In this context, `self.log_weight_scale` is not updated through `norm_factor`, but it is updated through `w2d`. In practical usage, should `.detach()` be removed from `self.log_weight_scale` to compute the loss using `norm_factor` for updating instead of relying on `w2d`?

**Comment by Einkazusa** [comment]
Thank you for your response again!
One more thing, when I train the network using EPro-PnP layer, a warning has appeare:
``` 
.conda\envs\EPro-PnP\lib\site-packages\torch\distributions\distribution.py:271: UserWarning: <class 'epropnp.distributions.AngularCentralGaussian'> does not define `support` to enable sample validation. Please initialize the distribution with `validate_args=False` to turn off validation.
  warnings.warn(f'{self.__class__} does not define `support` to enable ' +
```
  Will this affect the performance of the network?
